Develop

stepiko135 opened the pull request:
コードレビューよろしくお願いします！

Commit:
laravel/ui パッケージ導入

Commit:
MAMP環境でMySQLへの接続設定

Commit:
課題④-2

ReiNagahashi commented on laravelproject/routes/web.php:
```diff
@@ -20,3 +21,5 @@
 
 Route::get('/home', 'HomeController@index')->name('home');
+
+Route::resource('/boards', 'BoardController');
```
今回はメソッドshowを使用しないことから、思わぬエラーに備えてexceptを宣言しておいたほうが良いですね。
参考：https://qiita.com/sympe/items/9297f41d5f7a9d91aa11

stepiko135 replied:
ありがとうございました！改善しました！

ReiNagahashi commented on laravelproject/app/Http/Controllers/BoardController.php:
```diff
@@ -0,0 +1,108 @@
+    public function index()
+    {
+        $posts = Board::all();
+        return view('boards.index',['posts' => $posts]);
```
今回は返り値としてのプロパティは1つのみなので気にすることはありませんが、今後返すデータの数が多くなって行った場合はPHPメソッドのcompactを使用すると少ないコードで記述ができるようになるので、覚えておくといいですね。

stepiko135 replied:
compact便利ですね！
ありがとうございます！

stepiko135 committed:
修正
boards/index.blade.php　compactメソッドに変更、Eager Loadingに変更
BoardController.php　showメソッド削除

stepiko135 committed:
課題④−３

stepiko135 commented on laravelproject/resources/views/boards/index.blade.php:
```diff
@@ -21,9 +21,34 @@
+                <input type="hidden" value="{{$post->id}}" name='board_id'>
+                <input type="hidden" value="{{Auth::id()}}" name='user_id'>
+                <span class="material-icons ">
+                    @if(App\Like::where('board_id',$post->id)->where('user_id',Auth::id())->exists())
```
ページリロード後もいいねボタンの状態を維持するためにLikeController.php内でも記述した文を書いています。
もっと良い書き方はないかと思い質問します。

１，JavaScriptで実装することはできますか？一度挑戦しましたが、ｊｓの知識が浅く、ページリロード後も状態を維持する事ができませんでした。

２，テキストでは「コントローラー内の記述を増やすのは処理が遅くなる為避けるべき」とありましたが、そのためにテンプレート内の記述が増えてしまっても構わないということでしょうか？

３，これまでで一番苦労した課題だったので、わかりにくいコードかもしれません。更にきれいに書く方法があれば教えていただきたいです。

以上３点、宜しくおねがいします。

ReiNagahashi replied:
[1]Javascriptでも記述は可能ですが、今回のようにManyToManyのリレーションのデータが関係してきた場合は記述がやや複雑になってきますね。ですので今回はJSでの記述よりもLaravelのモデル・コントローラー名で実装をすることで、より簡単に確実にデータの作成・削除をすることが出来ます。

[2]やはり臨機応変に対応することが肝だと考えます。前回の課題３のようなフロントエンド上のboolenの要素を並べ替えるのみであればJSであれば素早く実装できますが、Laravelで実装すると非常に冗長になります。

[3]例えば、likesにあるpost_idを基にいいねの照会・作成・削除の実装は、いいねをするAuthユーザーからLikesテーブルにアクセスをすることでwhereを一度に複数使用することなく実装することが可能になります。

ReiNagahashi replied:
[3]の付け足しになりますが、実際にはAuth:user()->likes()->create([...])のように、カラムを作成するメソッドにアクセスすることが出来ます。

  ```
public function ownFavorite(Post $post){
        Auth::user()->favorites()->create([
            'post_id' => $post->id
        ]);
        return redirect(route('posts.index'));
    }

    public function deleteFavorite(Post $post){
        $favorite = Auth::user()->favorites()->where('post_id',$post->id)->first();
        $favorite->delete();
        return redirect(route('posts.index'));
    }
```

上記は私のコード上でのcreate, deleteになりますが、参考にしてみてください。

ReiNagahashi commented on laravelproject/app/User.php:
```diff
@@ -40,4 +40,8 @@ class User extends Authenticatable
         return $this->hasMany('App\Board');
     }
+
+    public function likes(){
```
前回の記述である、return $this->belongsToMany('App\Board','likes');が正常に動作しなかった原因はしばしばLaravel側がリレーションの記述の仕方によって認識してくれないことがあります。そんな場合には、今回のように確実にリレーションを組むことの出来る記述が出来ると良いですね。
